Auto-implement Interpolate and ComputeDistance for macro-generated keyword types

The keyword types are generated using [this macro](https://github.com/servo/servo/blob/46bb635c4c1eb9e420392d7a761956c1cc10ce3a/components/style/values/mod.rs#L111). The `Interpolate` and `ComputeDistance` traits are implemented for them in `components/style/properties/helpers/animated_properties.mako.rs`.

Since they're unit structs, these types don't have anything special, they always return their copy in `Interpolate` impl, and they always return an `Err(())` in `ComputeDistance` impl. So, we should do that with the help of the macro.

Alright, so I think I have a solution, but I'm not exactly sure what's going on, so I'm hoping to make this a learning experience. I removed the obvious `impl`'s for `Auto` and `Normal` and added the `impl` definitions inside the macro. I then started getting this build error:

<details>
<summary><b>Bugs in here</b></summary>

```
error[E0119]: conflicting implementations of trait `properties::animated_properties::Interpolate` for type `values::Either<app_units::Au, values::None_>`:
     --> /home/aaronc/Coding/rust/servo/target/debug/build/style-4f02e43a0ab74277/out/properties.rs:49595:1
      |
48545 | / impl Interpolate for LengthOrNone {
48546 | |     fn interpolate(&self, other: &Self, progress: f64) -> Result<Self, ()> {
48547 | |         match (*self, *other) {
48548 | |             (Either::First(ref length), Either::First(ref other)) =>
...     |
48552 | |     }
48553 | | }
      | |_- first implementation here
...
49595 | / impl<T, U> Interpolate for Either<T, U>
49596 | |         where T: Interpolate + Copy, U: Interpolate + Copy,
49597 | | {
49598 | |     #[inline]
...     |
49612 | |     }
49613 | | }
      | |_^ conflicting implementation for `values::Either<app_units::Au, values::None_>`
```
</details>

I also got this bug for the other `impl` I moved.

This was obviously fixed by removing the code for one of the `impl`'s, but I'm curious why the bug showed up. My idea is that by adding the `impl`'s to the macro, I added the two `impl`'s to `None_` that weren't there before. This caused the `where` clause of `impl Interpolate for Either` to match the same as `LengthOrNone`, since `LengthOrNone = Either<Length, None_>`.

So I'm not sure what the best path forward is. Should I leave it adding `Interpolate` and `ComputeDistance` to `None_` and remove one of the conflicting `impl`'s (my vote is removing `LengthOrNone`'s)? Or go back to how it was so that `None_` doesn't gain any new `impl`'s, since I'm not sure we want `None_` implementing those.

--- a/components/style/properties/helpers/animated_properties.mako.rs
+++ b/components/style/properties/helpers/animated_properties.mako.rs
@@ -33,9 +33,9 @@ use std::fmt;
 use style_traits::ToCss;
 use super::ComputedValues;
 use values::CSSFloat;
-use values::{Auto, Either, Normal, generics};
+use values::{Auto, Either, generics};
 use values::computed::{Angle, LengthOrPercentageOrAuto, LengthOrPercentageOrNone};
-use values::computed::{BorderRadiusSize, ClipRect, LengthOrNone};
+use values::computed::{BorderRadiusSize, ClipRect};
 use values::computed::{CalcLengthOrPercentage, Context, LengthOrPercentage};
 use values::computed::{MaxLength, MinLength};
 use values::computed::position::{HorizontalPosition, VerticalPosition};
@@ -628,20 +628,6 @@ impl Interpolate for Au {
     }
 }
 
-impl Interpolate for Auto {
-    #[inline]
-    fn interpolate(&self, _other: &Self, _progress: f64) -> Result<Self, ()> {
-        Ok(Auto)
-    }
-}
-
-impl Interpolate for Normal {
-    #[inline]
-    fn interpolate(&self, _other: &Self, _progress: f64) -> Result<Self, ()> {
-        Ok(Normal)
-    }
-}
-
 impl <T> Interpolate for Option<T>
     where T: Interpolate,
 {
@@ -1107,16 +1093,6 @@ impl Interpolate for ClipRect {
 
 ${impl_interpolate_for_shadow('BoxShadow', 'CSSParserColor::RGBA(RGBA::transparent())',)}
 ${impl_interpolate_for_shadow('TextShadow', 'CSSParserColor::RGBA(RGBA::transparent())',)}
-
-impl Interpolate for LengthOrNone {
-    fn interpolate(&self, other: &Self, progress: f64) -> Result<Self, ()> {
-        match (*self, *other) {
-            (Either::First(ref length), Either::First(ref other)) =>
-                length.interpolate(&other, progress).map(Either::First),
-            _ => Err(()),
-        }
-    }
-}
 
 /// Check if it's possible to do a direct numerical interpolation
 /// between these two transform lists.
@@ -2242,20 +2218,6 @@ impl ComputeDistance for Au {
     }
 }
 
-impl ComputeDistance for Auto {
-    #[inline]
-    fn compute_distance(&self, _other: &Self) -> Result<f64, ()> {
-        Err(())
-    }
-}
-
-impl ComputeDistance for Normal {
-    #[inline]
-    fn compute_distance(&self, _other: &Self) -> Result<f64, ()> {
-        Err(())
-    }
-}
-
 impl <T> ComputeDistance for Option<T>
     where T: ComputeDistance,
 {
@@ -2523,18 +2485,6 @@ impl ComputeDistance for LengthOrPercentageOrNone {
                 this.compute_distance(other)
             },
             _ => Err(())
-        }
-    }
-}
-
-impl ComputeDistance for LengthOrNone {
-    #[inline]
-    fn compute_distance(&self, other: &Self) -> Result<f64, ()> {
-        match (*self, *other) {
-            (Either::First(ref length), Either::First(ref other)) => {
-                length.compute_distance(other)
-            },
-            _ => Err(()),
         }
     }
 }

--- a/components/style/values/mod.rs
+++ b/components/style/values/mod.rs
@@ -11,6 +11,7 @@
 use Atom;
 pub use cssparser::{RGBA, Token, Parser, serialize_identifier, serialize_string};
 use parser::{Parse, ParserContext};
+use properties::animated_properties::{ComputeDistance, Interpolate};
 use std::ascii::AsciiExt;
 use std::borrow::Cow;
 use std::fmt::{self, Debug};
@@ -123,6 +124,20 @@ macro_rules! define_keyword_type {
         impl ::style_traits::ToCss for $name {
             fn to_css<W>(&self, dest: &mut W) -> ::std::fmt::Result where W: ::std::fmt::Write {
                 write!(dest, $css)
+            }
+        }
+
+        impl Interpolate for $name {
+            #[inline]
+            fn interpolate(&self, _other: &Self, _progress: f64) -> Result<Self, ()> {
+                Ok($name)
+            }
+        }
+
+        impl ComputeDistance for $name {
+            #[inline]
+            fn compute_distance(&self, _other: &Self) -> Result<f64, ()> {
+                Err(())
             }
         }
 

